### PR TITLE
Perf: remove Onyx usage from `withLocalize`

### LINKED ISSUE #4268: [Performance] Onyx usage in `withLocalize` adds a significant overhead
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?
_e.g. memory consumption, storage read/write times, React native bridge concerns, inefficient React component rendering, etc._
- storage read/write times
- inefficient React component rendering

This component is one of the most used components and has more than 310 instances depending on chat size
The current way that the HOC works is that it will make a new Onyx connection per each `withOnyx` instance. 
A benchmark reviewed that for a large chat the we're spending about 50sec for the `Onyx.get` related to retrieving the `preferredLocale` 

## What is the impact of this on end-users?
_List specific user experiences that will be improved by solving this problem e.g. app boot time, time to for some interaction to complete, etc._
- app boot time
- chat switching time

## List any benchmarks that show the severity of the issue
_Please also provide exact steps taken to collect metrics above if any so we can independently verify the results._

I'm using Onyx.printMetrics() to monitor the Onyx.get calls that happen during init: https://github.com/Expensify/react-native-onyx#benchmarks

#### Results

Device | Samsung Galaxy S7 Edge (2016)
-- | --
OS | Android 8.0

Total: | 123.80sec |   |   | Last | 17.49sec |   |   | ReportID | 71955477
-- | -- | -- | -- | -- | -- | -- | -- | -- | --

method | total time spent | max | min | avg | time last call completed | calls made
-- | -- | -- | -- | -- | -- | --
Onyx:get | 96.85sec | 2.92sec | 0.10ms | 144.77ms | 17.48sec | 669

- Full info in excel: https://docs.google.com/spreadsheets/d/1wGrVXsad6ayQZVqvyUmTuTb1cmeegdE0y-IFzhLiPKU/edit#gid=73551349&range=J8 (search for `preferredLocale`)

#### Setup
Using this Expensify/App hash: b2582f3a30a258639b3908d59989323c0f11aacb
Modify this file: `src/Expensify.js` 
- For `Onyx.init` add `captureMetrics: true`
- At the end of the file add `global.Onyx = Onyx;` 

Modify `src/libs/Navigation/AppNavigator/MainDrawerNavigator.js` 
- Make `getInitialReportScreenParams` always return the same report
- e.g. `return {reportID: '71955477'};` 

Start the app and connect with Flipper or a Hermes Debugger

#### Steps
1. Reload the app
2. Wait for processes to settle (about 5-10sec after LHN loads)
  - you can use `Onyx.printMetrics({ format: 'string' }})` when it starts to return the same value - everything is loaded
3. Run `Onyx.printMetrics({ format: 'csv', raw: true })`
4. Copy the results to a temp file
5. Import file as new sheet in excel	

## Proposed solution
Extract a `LocaleContextProvider` that wraps the `App` - the implementation would be pretty much identical to how the `class WithLocalize` works at the moment: https://github.com/Expensify/App/blob/ad773863b658ee45897b113b43028a2aaa14e69e/src/components/withLocalize.js#L49
The difference would be that it will `render` a `React` `ContextProvider` component 
Now only the top level component that wraps the `App` subscribes to Onyx
The HOC `withLocalize` changes to use a `LocaleContextConsumer` 
It uses the context to get translation related props and pass them to the `WrappedComponent` 
Everything else stays the same

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)
_Note: These should be the same as the benchmarks collected before any changes._
I've done a quick test to verify there will be a significant reduction in `Onyx.get` time. It's demonstrated in the initial changes in the PR here: #4253

Total: | 62.62sec |   |   | Last | 12.27sec |   |   | ReportID | 71955477
-- | -- | -- | -- | -- | -- | -- | -- | -- | --

method | total time spent | max | min | avg | time last call completed | calls made
-- | -- | -- | -- | -- | -- | --
Onyx:get | 40.44sec | 1.79sec | 0.08ms | 139.94ms | 12.27sec | 289

## Platform:
Where is this issue occurring?

- Web
- iOS
- Android (Biggest impact)
- Desktop App
- Mobile Web

**Version Number:** 1.0.80-2
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by kidroca**
@marcaaron 

### Details

This component is one of the most used components and has more than 310 instances depending on chat size
Instead of creating a new connection for each `withLocalize` usage - use the already cached value in `translate`
This significantly reduces the `Onyx.get` load that is caused by all the connections

### Fixed Issues (Related To)
Fixes #4268

### Tests

1. Measure performance (init time) before this change
2. Measure performance (init time) with the change in this PR

I'm using `Onyx.printMetrics()` to monitor the `Onyx.get` calls that happen during init: https://github.com/Expensify/react-native-onyx#benchmarks

##### Before
Total: | 123.80sec |   |   | Last | 17.49sec |   |   | ReportID | 71955477
-- | -- | -- | -- | -- | -- | -- | -- | -- | --

method | total time spent | max | min | avg | time last call completed | calls made
-- | -- | -- | -- | -- | -- | --
Onyx:get | 96.85sec | 2.92sec | 0.10ms | 144.77ms | 17.48sec | 669
- see the individual Onyx.get calls here: https://docs.google.com/spreadsheets/d/1wGrVXsad6ayQZVqvyUmTuTb1cmeegdE0y-IFzhLiPKU/edit?usp=sharing
  - search for `preferredLocale` 

##### After
Total: | 62.62sec |   |   | Last | 12.27sec |   |   | ReportID | 71955477
-- | -- | -- | -- | -- | -- | -- | -- | -- | --

method | total time spent | max | min | avg | time last call completed | calls made
-- | -- | -- | -- | -- | -- | --
Onyx:get | 40.44sec | 1.79sec | 0.08ms | 139.94ms | 12.27sec | 289
- excel data: https://docs.google.com/spreadsheets/d/1cVe3eEnwYgwbH4DRV__g9_W8kQFpVhvTjDu3Igr6Ds0/edit?usp=sharing

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

#### Web

image

### CODE REVIEW

**Review by iwiznia — DISMISSED** [review]
Nice improvement!!

**Comment by parasharrajat** [comment]
When we update the locale, Will the app re render components. I don't think it will.

File: src/libs/translate.js [hunk 1]
```diff
@@ -76,7 +76,12 @@ function translateLocal(phrase, variables) {
     return translate(preferredLocale, phrase, variables);
 }
 
+function getPreferredLocale() {
```
Comment (Julesssss) [inline src/libs/translate.js:79]: NAB: For consistency, we're missing a function doc here -- even though its function is obvious. 

Nice optimization!

**Comment by Julesssss** [comment]
> When we update the locale, Will the app re render components. I don't think it will.

That's a good point. I wouldn't have thought so either.

**Comment by iwiznia** [comment]
> When we update the locale, Will the app re render components. I don't think it will.

Ohhh... hmmm yeah, that kind of sucks.

**Comment by kidroca** [comment]
> When we update the locale, Will the app re render components. I don't think it will.

When does that happen, isn't the app reloading when a new locale is selected?

**Comment by iwiznia** [comment]
> When does that happen, isn't the app reloading when a new locale is selected?

You can test this by changing the language in the preferences (maybe you need to force being in the bet if that's not done by default in dev). And no, it does not reload, all components that are using withLocalize will re-render themselves.

**Comment by kidroca** [comment]
I've mentioned this at Slack, there's another way to implement the HOC and it will not suffer this problem: https://expensify.slack.com/archives/C01GTK53T8Q/p1627403864349200?thread_ts=1627393814.334100&cid=C01GTK53T8Q

The changes would be slightly more, but it might be an even faster if the HOC is implemented through `React.createContext`. As this will also save 300+ class instances of `withLocalize`

I'll try it out tomorrow and do another benchmark

File: src/libs/translate.js (hunk 1 above)
Comment (marcaaron) [reply src/libs/translate.js:79]: + If it has a return value it needs a doc but not necessarily a description.

**Comment by marcaaron** [comment]
Looks good so far and definitely up for more improvements. I'm adding a link back to the original issue in the description so there is some more context.

**Comment by marcaaron** [comment]
But I think maybe this needs a new issue?

**Comment by kidroca** [comment]
> But I think maybe this needs a new issue?

Yeah, I'll probably have to spent some more time here, and then test/bench the changes on the remaining platforms.

I can open an issue tomorrow

**Comment by marcaaron** [comment]
👍  Sounds great!

**Comment by kidroca** [comment]
@marcaaron I've created a new issue: #4268 and updated the PR description with the link

**Comment by kidroca** [comment]
Updated implementation and posted new benchmarks here: https://github.com/Expensify/App/issues/4268#issuecomment-890072411

I think another possible optimization that might be applied at the moment is to `drop` the `withOnyx` call altogether like this:

```jsx
    componentDidMount() {
        this.connectionId = Onyx.connect({
            key: ONYXKEYS.NVP_PREFERRED_LOCALE,
            callback: (val) => {
                if (val) {
                    this.setState({preferredLocale: val});
                }
            },
        });
    }

    componentWillUnmount() {
        Onyx.disconnect(this.connectionId);
    }
```

This skips creating the `withOnyx` class component
I've tried it briefly it was slightly faster bringing "Last call ended time" to ~10.87s (otherwise it's ~11.05s) 
I can try to capture more data about it next week, but it might be better to focus/find something more important

File: src/components/withLocalize.js [hunk 2]
```diff
@@ -29,124 +31,122 @@ const withLocalizePropTypes = {
...
+    },
+})(LocaleContextProvider);
+
+Provider.displayName = 'withOnyx(LocaleContextProvider)';
+
+export default function withLocalize(WrappedComponent) {
+    const WithLocalize = forwardRef((props, ref) => (
+        <LocaleContext.Consumer>
+            {/* eslint-disable-next-line react/jsx-props-no-spreading */}
+            { translateUtils => <WrappedComponent {...translateUtils} {...props} ref={ref} />}
```
Comment (marcaaron) [inline src/components/withLocalize.js:140]: NAB, Unnecessary space here.

**Review by marcaaron — APPROVED** [review]
These changes look good to me and do make chat switching significantly faster. I can't really think of any reasons why we should not do this (other than people being unfamiliar with the Context API). Probably as a follow up we can add something to the documentation about how to approach using `withOnyx` when a component will be used a great number of times throughout the app or some list.

**Comment by marcaaron** [comment]
>This skips creating the withOnyx class component
>I've tried it briefly it was slightly faster bringing "Last call ended time" to ~10.87s (otherwise it's ~11.05s)
>I can try to capture more data about it next week, but it might be better to focus/find something more important

Yeah, I think there are probably better optimizations to focus on. This is already so great 🎉

**Comment by isagoico** [comment]
@kidroca Hello! Any QA tests needed for this PR?

**Comment by kidroca** [comment]
@isagoico No specific test steps other than basic regressions as translations are used everywhere. 
Switching the app language should work

**Comment by nkuoch** [comment]
@kidroca this PR created [this regression](https://github.com/Expensify/App/issues/4352#issuecomment-890863100).
Seems like [it was pointed out](https://github.com/Expensify/App/pull/4253#issuecomment-887555934) before the merge though. Not sure why we merged it regardless and what was decided about it though? cc @marcaaron

**Comment by iwiznia** [comment]
> Are we going to handle this as a continuation PR that fixes the regression?

IMO yes, because it's a feature that worked and was broken by this PR.

**Comment by parasharrajat** [comment]
@kidroca Please close that Issue when you submit a fix via the PR.

**Comment by kidroca** [comment]
Shouldn't I be assigned to the new ticket as well then?

**Comment by parasharrajat** [comment]
Is that any advantage of assigning an issue? Asking in general.

**Comment by iwiznia** [comment]
It makes it easier to find things if they are assigned to the people that worked on them. It is also used in some of our internal tooling to for example assign reviewers for issues or PRs.

**Comment by marcaaron** [comment]
>Not sure why we merged it regardless and what was decided about it though? 
 
Oh, hmm, we merged it because the implementation was entirely different. In short, we didn't suspect the new one would have the same issue.

**Pull request merged**